src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+            _frontendReadBackendWriteTask = null;
+        }
+
+        public bool IsRunning {
+            get
+            {
+                if (_backendReadFrontendWriteTask == null || _frontendReadBackendWriteTask == null || _disposed)
+                    return false;
+
+                return !_backendReadFrontendWriteTask.IsCompleted && !_frontendReadBackendWriteTask.IsCompleted;
+            }
+        }
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                Stop();
+
+                _cancelRouterTokenSource.Dispose();
+
+                _backendStream?.Dispose();
+                _frontendStream?.Dispose();
+
+                _disposed = true;
+
+                Interlocked.Decrement(ref s_routerInstanceCount);
+
+                _logger.LogDebug($"Diposed stats: Backend->Frontend {_backendToFrontendByteTransfer} bytes, Frontend->Backend {_frontendToBackendByteTransfer} bytes.");
+                _logger.LogDebug($"Active instances: {s_routerInstanceCount}");
+            }
+        }
+
+        async Task BackendReadFrontendWrite(CancellationToken token)
+        {
+            try
+            {
+                byte[] buffer = new byte[1024];
+                while (!token.IsCancellationRequested)
+                {
+                    _logger.LogDebug("Start reading bytes from backend.");
+
+                    int bytesRead = await _backendStream.ReadAsync(buffer, 0, buffer.Length, token).ConfigureAwait(false);
+
+                    _logger.LogDebug($"Read {bytesRead} bytes from backend.");
+
+                    // Check for end of stream indicating that remote end hung-up.
+                    if (bytesRead == 0)
+                    {
+                        _logger.LogDebug("Backend hung up.");
+                        break;
+                    }
+
+                    _backendToFrontendByteTransfer += (ulong)bytesRead;
+
+                    _logger.LogDebug($"Start writing {bytesRead} bytes to frontend.");
+
+                    await _frontendStream.WriteAsync(buffer, 0, bytesRead, token).ConfigureAwait(false);
+                    await _frontendStream.FlushAsync().ConfigureAwait(false);
+
+                    _logger.LogDebug($"Wrote {bytesRead} bytes to frontend.");
+                }
+            }
+            catch (Exception)
+            {
+                // Completing task will trigger dispose of instance and cleanup.
+                // Faliure mainly consists of closed/disposed streams and cancelation requests.
+                // Just make sure task gets complete, nothing more needs to be in response to these exceptions.
+                _logger.LogDebug("Failed stream operation. Completing task.");
+            }
+
+            RouterTaskCompleted?.TrySetResult(true);
+        }
+
+        async Task FrontendReadBackendWrite(CancellationToken token)
+        {
+            try
+            {
+                byte[] buffer = new byte[1024];
```

Comment:
same as above

Reply:
This allocation is already reused for the complete router session.